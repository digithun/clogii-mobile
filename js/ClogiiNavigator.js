import React from 'react';
import {
  Navigator,
  Linking
} from 'react-native';
import {parse} from 'query-string';

import ClogiiTabView from './tabs/ClogiiTabView';
import Player from './player';
import Book from './book';

class ClogiiNavigator extends React.Component {
  constructor(...args) {
    super(...args);
    this.onOpenURL = this.onOpenURL.bind(this);
  }

  componentDidMount() {
    Linking.getInitialURL().then((url) => {
      if (url) {
        this.fromURL(url);
      }
    }).catch(error => {
      console.error('initial url error: ', error);
    });
    Linking.addEventListener('url', this.onOpenURL);
  }

  componentWillUnmount() {
    Linking.removeEventListener('url', this.onOpenURL);
  }

  render() {
    return (
      <Navigator
        ref="navigator"
        initialRoute={{page: 'main-tab', id: 3}}
        renderScene={this.renderScene.bind(this)}
        />
    );
  }

  onOpenURL(event) {
    this.fromURL(event.url);
  }

  fromURL(url) {
    let {method, query} = this.parseURL(url);
    if (method === 'player' && query.id) {
      this.refs.navigator.push({page: 'player', id: query.id});
    }
  }

  parseURL(url) {
    try {
      url = url.split(':')[1];
      let [method, queryS] = url.split('?');
      let query = parse(queryS);
      if (method.indexOf('//') === 0) {
        method = method.slice(2);
      }
      return {method, query};
    }
    catch (error) {
      console.error('parse url error: ', error);
    }
    return {};
  }

  renderScene(route, navigator) {
    if (route.page === 'main-tab') {
      return <ClogiiTabView navigator={navigator}/>;
    }
    if (route.page === 'player') {
      return <Player onBackPress={this.onBackPress.bind(this)} navigator={navigator} id={route.id}/>;
    }
    if (route.page === 'book') {
      return <Book
        onBackPress={this.onBackPress.bind(this)}
        goToPlayer={this.goToPlayer.bind(this)}
        id={route.id}/>;
    }
  }

  goToPlayer(id) {
    this.refs.navigator.push({page: 'player', id})
  }

  onBackPress() {
    this.refs.navigator.pop();
  }
}

export default ClogiiNavigator;
