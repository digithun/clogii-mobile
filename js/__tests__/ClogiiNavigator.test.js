import React from 'react';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import toJSON from 'enzyme-to-json';

import {View} from 'react-native';

import ClogiiNavigator from '../ClogiiNavigator';

function getWrapper(props = {}) {
    return shallow(<ClogiiNavigator {...props}/>);
}

describe('ClogiiNavigator', () => {
    it('render', () => {
        const wrapper = getWrapper();
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('navigate to Book/Player', () => {
        const dump = getWrapper();
        const spy = jest.fn();
        dump.instance().refs = {
            navigator: {
                push: spy
            }
        };
        dump.instance().goToPlayer(5);
        expect(spy).toBeCalledWith({id: 5, page: "player"});
        dump.instance().goToBook(123);
        expect(spy).toBeCalledWith({id: 123, page: "book"});
    });

    it('navigate back', () => {
        const dump = getWrapper();
        const spy = jest.fn();
        dump.instance().refs = {
            navigator: {
                pop: spy
            }
        };
        dump.instance().goBack(5);
        expect(spy).toBeCalled();
    });

    it('onOpenURL will map event to fromURL', () => {
        const dump = getWrapper();
        const spy = jest.fn();
        dump.instance().fromURL = spy;
        dump.instance().onOpenURL({url: 'x'});
        expect(spy).toBeCalledWith('x');
    });

    it('fromURL will navigate to player on `clogii://player?id=123`', () => {
        const dump = getWrapper();
        const spy = jest.fn();
        dump.instance().refs = {
            navigator: {
                push: spy
            }
        };
        dump.instance().fromURL(`clogii://player?id=123`);
        expect(spy).toBeCalledWith({page: 'player', id: '123'});
    });

    it('parseURL split to method and query', () => {
        const dump = getWrapper();
        const result = dump.instance().parseURL('clogii://player?id=199');
        expect(result.method).toBe('player');
        expect(result.query).toEqual({id: "199"});

        const result2 = dump.instance().parseURL('clogii://book?id=77&x=3');
        expect(result2.method).toBe('book');
        expect(result2.query).toEqual({id: "77", x: "3"});
    });

    describe('renderScene', () => {
        it('main-tab', () => {
            const dump = getWrapper();
            const tree = shallow(<View>{dump.instance().renderScene({page: 'main-tab'})}</View>);
            expect(toJSON(tree)).toMatchSnapshot();
        });

        it('player', () => {
            const dump = getWrapper();
            const tree = shallow(<View>{dump.instance().renderScene({page: 'player', id: 1})}</View>);
            expect(toJSON(tree)).toMatchSnapshot();
        });

        it('book', () => {
            const dump = getWrapper();
            const tree = shallow(<View>{dump.instance().renderScene({page: 'book', id: 3})}</View>);
            expect(toJSON(tree)).toMatchSnapshot();
        });
    });
});