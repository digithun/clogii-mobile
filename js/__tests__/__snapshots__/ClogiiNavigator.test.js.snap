exports[`ClogiiNavigator render 1`] = `
<FixBugScrollViewNavigator
  initialRoute={
    Object {
      "id": 3,
      "page": "main-tab",
    }
  }
  renderScene={[Function]} />
`;

exports[`ClogiiNavigator renderScene book 1`] = `
<View>
  <Apollo(Book)
    goToPlayer={[Function]}
    id={3}
    onBackPress={[Function]} />
</View>
`;

exports[`ClogiiNavigator renderScene main-tab 1`] = `
<View>
  <Connect(ClogiiTabView)
    goToBook={[Function]} />
</View>
`;

exports[`ClogiiNavigator renderScene player 1`] = `
<View>
  <Apollo(Player)
    id={1}
    onBackPress={[Function]} />
</View>
`;
